guidelines/design-guidelines.md

```diff
@@ -31,4 +46,148 @@ Example:
+   - Examples: `NServiceBus.Gateway`, `NServiceBus.Persistence.AzureStorage` etc.
+* For public types designed for extensibility, we use the root [`{Component}`](#Component naming rules) namespace. This hides types irrelevant for business developers while making them discoverable for developers extending NServiceBus. 
+
+
+## Component naming rules
+
+Components can be split up into two types: those that are unique and those that belongs to a category.
+
+Unique components should be named appropriately. E.g. `NServiceBus.Gateway`, `NServiceBus.Callbacks`, etc.
+
+Components belonging to a category should be named `NServiceBus.{Category}.*`. E.g. `NServiceBus.Persistence.AzureStorage`, `NServiceBus.Persistence.MongoDb`, etc.
+
+## Composition
+
+The public API of NServiceBus is a composition API consisting of multiple capabilities. The APIs in NServiceBus are extensible enough so that [capabilities](https://github.com/Particular/Vision/labels/Capability) can extend the composition API without needing to share the same release cycle as NServiceBus, reducing the coupling and allowing us to organize code in a more cohesive way.
+
+In order to achieve this goal, a few key design decision have been made:
+
+### Folders to group capabilities
+
+Folders are used to group source files together to represent individual capabilities without affecting the namespaces of the components grouped together in these folders.
+
+For example, NServiceBus has a Recoverability capability folder, which is further divided into:
+
+* Faults
+* ImmediateRetries
+* DelayedRetries
+
+### Extension points on public APIs
+
+Public APIs provide extension points, allowing capabilities to hook in their business logic and float required state over those APIs to various extensions.
+
+For example, `SendOptions`, `PublishOptions` and `ReplyOptions` inherit from `ExtendableOptions` which provides a `ContextBag` to float additional state into the option classes. This state is then made available on the outgoing pipeline.
+
+### State belongs to a specific capability
+
+Instead of directly attaching state to composition roots and thereby exposing that state to other capabilities, we instead use the extension APIs to keep state internal to the capability.
+
+For example, let's consider `SendOptions` and `SendLocal`. We do not expose a `RouteToThisEndpoint` property on `SendOptions`:
+
+```
+class SendOptions
+  public bool RouteToThisInstance { get; set; }
+```
+
+Instead, an extension method called `RouteToThisInstance` is used to set the internal state. The internal state belongs to the routing capability. Therefore the state cannot be attached to `SendOptions`. Since C#/.NET doesn't support extension properties, the only way to implement this is to use an extension method. The benefit of extension methods is that the reader and the writer of an option of a capability can be implemented with different names.
```

Comment:
> The benefit of extension methods is that...

The benefit over what? It's not clear that it's methods over properties. I previously suggested:

> The benefit of extension methods *, as opposed to properties,* is that...

But a simpler fix could be:

> The benefit of ~extension~ methods *over properties* is that...